go-controller/pkg/node/gateway_localnet.go

```diff
@@ -151,92 +232,185 @@ func initLocalnetGateway(nodeName string, subnet *net.IPNet, wf *factory.WatchFa
+	for _, subnet := range subnets {
+		if utilnet.IsIPv6CIDR(subnet) {
+			ipV6NetworkCount++
+		} else {
+			ipV4NetworkCount++
+		}
+	}
+	if ipV4NetworkCount > supportedSubnetsPerFamily || ipV6NetworkCount > supportedSubnetsPerFamily {
+		return fmt.Errorf("Only one IPv4 and/or IPv6 subnets can be specified. Specified %d IP v4 and %d "+
+			"IP v6 subnets", ipV4NetworkCount, ipV6NetworkCount)
 	}
+	return nil
+}
 
-	err = localnetGatewayNAT(ipt, localnetGatewayNextHopPort, gatewayIP)
-	if err != nil {
-		return fmt.Errorf("Failed to add NAT rules for localnet gateway (%v)", err)
+func constructIPData(subnets []*net.IPNet) ([]*localnetData, error) {
+	netData := make([]*localnetData, 0)
+	for _, subnet := range subnets {
+		if utilnet.IsIPv6CIDR(subnet) {
+			data, err := newIPV6Data()
+			if err != nil {
+				return nil, err
+			}
+			netData = append(netData, data)
+		} else {
+			data, err := newIPV4Data()
+			if err != nil {
+				return nil, err
+			}
+			netData = append(netData, data)
+		}
 	}
+	return netData, nil
+}
 
-	if config.Gateway.NodeportEnable {
-		err = localnetNodePortWatcher(ipt, wf, gatewayIP)
+func newIPV4Data() (*localnetData, error) {
+
+	ipt, err := util.GetIPTablesHelper(iptables.ProtocolIPv4)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize iptables: %v", err)
 	}
+	gatewayIP := net.ParseIP(v4localnetGatewayIP)
+	gatewayNextHop := net.ParseIP(v4localnetGatewayNextHop)
+	gatewaySubnetMask := net.CIDRMask(v4localnetGatewaySubnetPrefix, 32)
 
-	return err
+	data := newIPData(ipv4, ipt, gatewayIP, gatewayNextHop, gatewaySubnetMask)
+	return data, nil
 }
 
-// localnetIPTablesHelper gets an IPTablesHelper for IPv4 or IPv6 as appropriate
-func localnetIPTablesHelper(subnet *net.IPNet) (util.IPTablesHelper, error) {
-	var ipt util.IPTablesHelper
-	var err error
-	if utilnet.IsIPv6CIDR(subnet) {
-		ipt, err = util.GetIPTablesHelper(iptables.ProtocolIPv6)
-	} else {
-		ipt, err = util.GetIPTablesHelper(iptables.ProtocolIPv4)
-	}
+func newIPV6Data() (*localnetData, error) {
+
+	ipt, err := util.GetIPTablesHelper(iptables.ProtocolIPv6)
 	if err != nil {
 		return nil, fmt.Errorf("failed to initialize iptables: %v", err)
 	}
-	return ipt, nil
+	gatewayIP := net.ParseIP(v6localnetGatewayIP)
+	gatewayNextHop := net.ParseIP(v6localnetGatewayNextHop)
+	gatewaySubnetMask := net.CIDRMask(v6localnetGatewaySubnetPrefix, 128)
+
+	data := newIPData(ipv6, ipt, gatewayIP, gatewayNextHop, gatewaySubnetMask)
+	return data, nil
+}
+
+func newIPData(iPVersion int, ipt util.IPTablesHelper, gatewayIP net.IP, gatewayNextHop net.IP, gatewaySubnetMask net.IPMask) *localnetData {
+	data := &localnetData{
+		ipVersion:          iPVersion,
+		ipt:                ipt,
+		gatewayIP:          gatewayIP,
+		gatewayNextHop:     gatewayNextHop,
+		gatewaySubnetMask:  gatewaySubnetMask,
+		gatewayIPCIDR:      &net.IPNet{IP: gatewayIP, Mask: gatewaySubnetMask},
+		gatewayNextHopCIDR: &net.IPNet{IP: gatewayNextHop, Mask: gatewaySubnetMask},
```

Comment:
`gatewayIPCIDR` and `gatewayNextHopCIDR` are each only used in one place, so they could be removed from `localnetData` and you could just put the `&net.IPNet{...}` definition inline where they're needed

Reply:
Will look into this.